### PR TITLE
EAMxx: Option to run SHOC with no SGS variability (1.5 closure)

### PR DESCRIPTION
**Pull request by bogensch**
This PR adds an option to run SHOC with no SGS variability.  This essentially reduces SHOC to a 1.5 TKE closure in the vertical.

If this option is activated (shoc_1p5tke= true):
- The scalar variances and covariances are set to zero.
- The third moment of vertical velocity is set to zero.
- The above two assumptions will reduce the assumed PDF to an all-or-nothing closure.
- Since an all-or-nothing closure is used this means SHOC’s buoyancy flux parameterization is now ill-posed due to the liquid water flux (w'ql') always being zero.  The buoyancy flux is needed to close the buoyant production term in the TKE budget.  In this case we close the buoyant production of TKE using the local moist brunt vaisalla frequency, which follows many 1.5 TKE closures.
- Modifies the definition of the eddy diffusivities to align better with formulations used in a 1.5 scheme.
- Modifies the length scale definition to be exactly that of SAM when a 1.5 TKE closure is activated.

I have verified in several cases, and one ne30 simulation, that if this option is activate that cloud fraction will always be either 0 or 1 using instantaneous output.

Since this scheme is set to false by default, this should be a b4b PR.

### CODE REVIEW

File: components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.cpp [hunk 1]
```diff
@@ -1813,7 +1818,7 @@ void diag_third_shoc_moments_host(Int shcol, Int nlev, Int nlevi, Real* w_sec, R
 }
 
 void adv_sgs_tke_host(Int nlev, Int shcol, Real dtime, Real* shoc_mix, Real* wthv_sec,
```
[the remark this answers is not in the document]
Comment (AaronDonahue) [reply components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.cpp:1820]: I suppose not, it looks like this function is only for testing.  Does that mean we are not testing the `shoc_nosgs_var=.true.` option?

**Review by AaronDonahue — COMMENTED** [review]
Looks good to me.  I had a question about one of the functions.

File: components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bogensch) [reply components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.cpp:1820]: I KNEW you were going to ask me this.  So, the “meat” of the no SGS variability concept is already property tested in the shoc_assumed_pdf_tests.cpp tests.  Here the very first tests ensure that if the input scalar variances are zero that the scheme will reduce to an all-or-nothing closure.  This is by far the most important test and it is already covered.

I did not add tests for the set-up of the all-or-nothing scheme which entails just setting several variables to zero in a couple routines if shoc_no_sgs_var is activated.  I suppose you are going to make me do this, though?  ;)

[the remark this answers is not in the document]
Comment (AaronDonahue) [reply components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.cpp:1820]: How painful would it be to create a test for the setup?  I know it seems trivial, but the spirit of testing is that we even cover trivial cases to be certain everything works as expected.

[the remark this answers is not in the document]
Comment (bogensch) [reply components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.cpp:1820]: understood.  I already conceded to the fact that I would have to do this, plus it will give me more experience.  Will add some tests for the routines I modified.

**Comment by tcclevenger** [comment]
> Setting to WIP while I rework a few things.

I think the preferred method now is to set PR as "Draft". Testing will still run, but you won't be questioned about the progress.

**Comment by bogensch** [comment]
This PR is now open again for your re-reviewing pleasure.  It is much more expansive than the first go around and should probably be scrutinized closely since I’m not 100% confident I hooked up the boiler plate to the property tests properly.  For instance, the components/eamxx/src/physics/shoc/tests/infra/shoc_test_data.hpp file is throwing up warnings in the Github viewer; which seems to be related to me adding the bool in the data types.  With that said, all shoc tests are running and passing on my end.

File: components/eamxx/src/physics/shoc/eamxx_shoc_process_interface.cpp [hunk 2]
```diff
@@ -258,6 +258,7 @@ void SHOCMacrophysics::initialize_impl (const RunType run_type)
   runtime_options.c_diag_3rd_mom = m_params.get<double>("c_diag_3rd_mom");
   runtime_options.Ckh           = m_params.get<double>("coeff_kh");
   runtime_options.Ckm           = m_params.get<double>("coeff_km");
+  runtime_options.shoc_1p5tke = m_params.get<bool>("shoc_1p5tke");
```
[the remark this answers is not in the document]
Comment (tcclevenger) [reply components/eamxx/src/physics/shoc/eamxx_shoc_process_interface.cpp:261]: Might be beneficial for all these params... (but not needed for this PR)

**Review by AaronDonahue — COMMENTED** [review]
I have nothing to add on top of the comments from Naser and Conrad.  Once they are satisfied then I am happy to approve as well.

One science question, would we ever expect someone to want to set the closure to a value other than 1.5?  I'm noting that here all of the variables are names with the closure value implied.  We don't need to fix anything now, we can change the variable names if this every becomes a runtime option, but just asking as it came to mind while I was reviewing the PR.

File: components/eamxx/src/physics/shoc/eamxx_shoc_process_interface.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bogensch) [reply components/eamxx/src/physics/shoc/eamxx_shoc_process_interface.cpp:261]: Yeah, I left as is for this PR to follow precedent for the other params.

File: components/eamxx/src/physics/shoc/impl/shoc_diag_second_moments_impl.hpp [hunk 3]
```diff
@@ -67,6 +67,24 @@ void Functions<S,D>::diag_second_moments(
...
+    const Int nlev_pack = ekat::npack<Spack>(nlev);
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_pack), [&] (const Int& k) {
+      w_sec(k) = 0;
+    });
+
+    const Int nlevi_pack = ekat::npack<Spack>(nlevi);
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlevi_pack), [&] (const Int& k) {
+      thl_sec(k) = 0;
+      qw_sec(k) = 0;
+      qwthl_sec(k) = 0;
```
[the remark this answers is not in the document]
Comment (bogensch) [reply components/eamxx/src/physics/shoc/impl/shoc_diag_second_moments_impl.hpp]: Thanks, I fixed these blocks of code to follow your alls suggestions.

File: components/eamxx/src/physics/shoc/impl/shoc_compute_shoc_mix_shoc_length_impl.hpp [hunk 4]
```diff
@@ -30,10 +33,29 @@ ::compute_shoc_mix_shoc_length(
...
+	// Search for stable cells
+        const auto stable_mask = brunt(k) > 0;
+
+        // Define length scale for stable cells
+        const auto length_tmp = ekat::sqrt(0.76*tk(k)/0.1/ekat::sqrt(brunt(k) + 1.e-10));
+        // Limit the stability corrected length scale between 0.1*dz and dz
+	const auto limited_len = ekat::min(dz_zt(k),ekat::max(0.1*dz_zt(k),length_tmp));
+
+        // Overwrite the length scale in stable cells with the new definition
+        shoc_mix(k).set(stable_mask, limited_len);
```
[the remark this answers is not in the document]
Comment (bogensch) [reply components/eamxx/src/physics/shoc/impl/shoc_compute_shoc_mix_shoc_length_impl.hpp]: I revised accordingly.

**Comment by bogensch** [comment]
@AaronDonahue I assume you're asking whether a user could specify a different closure value, such as 2.0 or 3.0? This is not currently supported, and I don't expect it to be at any point in the future. Adjusting SHOC to any closure other than 1.5 is nontrivial and would require extensive code modifications and additions.  At that point it would make sense to just make a completely new parameterization.

**Review by tcclevenger — APPROVED** [review]
This looks good to me!

**Review by AaronDonahue — APPROVED** [review]

**Comment by tcclevenger** [comment]
Shoc unit tests failing on both gpu and cpu. CIME cases seem fine.

**Comment by AaronDonahue** [comment]
@tcclevenger would you be able to check on the test that Peter B. posts about up above?

**Comment by tcclevenger** [comment]
@bogensch @AaronDonahue Yeah, unit test related, should be a simple fix. I'll take a look.

File: components/eamxx/src/physics/shoc/tests/shoc_diag_second_moments_tests.cpp [hunk 5]
```diff
@@ -247,17 +250,44 @@ struct UnitWrap::UnitTest<D>::TestDiagSecondMoments : public UnitWrap::UnitTest<
...
 
     DiagSecondMomentsData baseline_data[] = {
-      DiagSecondMomentsData(36,  72, 73),
-      DiagSecondMomentsData(72,  72, 73),
-      DiagSecondMomentsData(128, 72, 73),
-      DiagSecondMomentsData(256, 72, 73),
+      DiagSecondMomentsData(36,  72, 73, 2),
+      DiagSecondMomentsData(72,  72, 73, 2),
+      DiagSecondMomentsData(128, 72, 73, 2),
+      DiagSecondMomentsData(256, 72, 73, 2),
```
Comment (tcclevenger) [inline components/eamxx/src/physics/shoc/tests/shoc_diag_second_moments_tests.cpp:290]: @bogensch Isn't this a bool? Why is it being set to `2`?

**Comment by bogensch** [comment]
Thanks @tcclevenger for your help.  This issue should be addressed now.

**Comment by tcclevenger** [comment]
So I have triggered the fails on my machine, and they are purely due to adding a data input in the shoc unit tests, so loading baselines complains that the size of the file data doesn't match the test data struct (which is expected). We can then merge and bless baselines and we are good.

The issue is that it could be that this PR is non-BFB, but because we fail before comparing baselines we miss it, and we bless non-bfb changes. So then the thing to do would be remove `shoc_1p5tke` from the testing data structs, add it manually to the `_host()` test functions, verify BFB, merge, then new PR which moves that to the testing structs. 

That's way overkill in my opinion because all the CIME cases are passing, so I vote that I merge and bless. I'll do this if I can get @AaronDonahue or @bartgol to agree. (Or they can point out a flaw in my logic)

**Comment by tcclevenger** [comment]
Simply commenting out the check for file size being equal doesn't allow me to test. Next week I plan to manually set this new param outside of the BFB tests so that tests will pass. Later we can think about passing it to the test data structures if we want to test both true and false.

**Comment by bartgol** [comment]
@tcclevenger if you are sure that the failure is due to the new param (that is not found in the baselines file), then I would just merge and bless.

> Simply commenting out the check for file size being equal doesn't allow me to test.

I am not sure I understand this comment though. What happens when you comment the check?

**Comment by tcclevenger** [comment]
> @tcclevenger if you are sure that the failure is due to the new param (that is not found in the baselines file), then I would just merge and bless.

I'm not sure, I would have guessed that it was due to the new param, but it's not clear to me that is the case. So I want to do at least a quick check.

When I comment out the check I get a baseline diff which contains wildly different values. This doesn't make sense to me since the CIME cases are BFB. Not sure if it's unit test specific, or if the files being mismatched sizes means that we are comparing the correct values.

**Comment by tcclevenger** [comment]
@mahf708 Manually setting new param to `false` and removing from the testing data struct causes both the file size comparison and the bfb array comparisons to pass. I think the errors come from comparing files of different sizes, and not BFB issues in the source code. I'm going to test with the 2 examples you posted to be sure. 

If we want to go a more rigorous route, I could make a quick PR that only adds the new param to the testing data, merge that and bless baselines, then have this PR actually use the new param.

**Comment by tcclevenger** [comment]
Yeah, tested the examples and same thing, I'm confident this BFB. Merging and blessing.

**Comment by tcclevenger** [comment]
> Looks like this PR caused some fpe errors: https://my.cdash.org/tests/273678509

Investigating.
